fuel-p2p/src/orchestrator.rs

```diff
@@ -0,0 +1,140 @@
+    tx_transaction: Sender<TransactionBroadcast>,
+    tx_block: Sender<BlockBroadcast>,
+
+    db: Arc<dyn RelayerDb>,
+
+    outbound_responses: FuturesUnordered<ResponseFuture>,
+}
+
+impl NetworkOrchestrator {
+    pub async fn new(
+        local_keypair: Keypair,
+        p2p_config: P2PConfig,
+        rx_request_event: Receiver<P2pRequestEvent>,
+
+        tx_consensus: Sender<ConsensusBroadcast>,
+        tx_transaction: Sender<TransactionBroadcast>,
+        tx_block: Sender<BlockBroadcast>,
+
+        db: Arc<dyn RelayerDb>,
+    ) -> Self {
+        let p2p_service = FuelP2PService::new(local_keypair, p2p_config)
+            .await
+            .unwrap();
+
+        Self {
+            p2p_service,
+            rx_request_event,
+            tx_block,
+            tx_consensus,
+            tx_transaction,
+            db,
+            outbound_responses: Default::default(),
+        }
+    }
+
+    pub async fn run(&mut self) {
+        loop {
+            tokio::select! {
+                next_response = self.outbound_responses.next() => {
+                    if let Some(Some((response, request_id))) = next_response {
+                        let _ = self.p2p_service.send_response_msg(request_id, response);
+                    }
+
+                },
+                p2p_event = self.p2p_service.next_event() => {
+                    if let FuelP2PEvent::Behaviour(behaviour_event) = p2p_event {
+                        match behaviour_event {
+                            FuelBehaviourEvent::GossipsubMessage { message, .. } => {
+                                match message {
+                                    GossipsubMessage::NewTx(tx) => {
+                                        let _ = self.tx_transaction.send(TransactionBroadcast::NewTransaction(tx));
+                                    },
+                                    GossipsubMessage::NewBlock(block) => {
+                                        let _ = self.tx_block.send(BlockBroadcast::NewBlock(block));
+                                    },
+                                    GossipsubMessage::ConensusVote(vote) => {
+                                        let _ = self.tx_consensus.send(ConsensusBroadcast::NewVote(vote));
+                                    },
+                                }
+                            },
+                            FuelBehaviourEvent::RequestMessage { request_message, request_id } => {
+                                match request_message {
+                                    RequestMessage::RequestBlock(block_height) => {
+                                        let db = self.db.clone();
+
+                                        self.outbound_responses.push(
+                                            Box::pin(async move {
+                                                db.get_sealed_block(block_height).await.map(|block| (OutboundResponse::ResponseBlock(block), request_id))
```

Comment:
It may make sense to have some other service/module handle fetching blocks from the DB for the orchestrator. For instance, if there are many outstanding requests (e.g. lots of full-syncs) having a dedicated block-fetching worker on the [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) threadpool would help avoid stalling the main async runtime.

Reply:
since the fetching itself is already async, shouldn't it be just fine to spawn another async task with `tokio::spawn` instead ?

Reply:
good point, it should be the responsibility of this async-trait implementor to deal with any potential blocking issues.